# Importing my data

Data are imported from this Kaggle: https://www.kaggle.com/datasets/threnjen/board-games-database-from-boardgamegeek

List of my dataframes (9):
games, mechanics, themes, subcategories, artists_reduced, designers_reduced, publishers_reduced, user_ratings, ratings_distribution

For more info regarding the columns of each, please look at the 'bgg_data_documentation' text file

## <b>Importing </b>

In [1]:
import pandas as pd

In [2]:
# Location where my CSV files are stored:
file_path = r'D:\Docs Persos\IRONHACK_BC\FINAL PROJECT\Possible Datasets\BGG - Board Game Geek\Board Game Database from BoardGameGeek'

## Reading the various datasets

This area is for the following:
- Reading the different dataframes <br>
- Pivoting + reseting index <br>
- Creation of new_ones based on the biggest one

### <b> Games </b>

The main df of the dataset - we'll modify and clean it later

In [3]:
# Importing the main dataframe 'games.csv'
df_games = pd.read_csv(file_path + '/games.csv')

# Display the dataframe:
display(df_games.head(3))

# looking at the size of the df
print(df_games.shape)

BGGId          Name                                        Description  \
0      1    Die Macher  die macher game seven sequential political rac...   
1      2  Dragonmaster  dragonmaster tricktaking card game base old ga...   
2      3       Samurai  samurai set medieval japan player compete gain...   

   YearPublished  GameWeight  AvgRating  BayesAvgRating   StdDev  MinPlayers  \
0           1986      4.3206    7.61428         7.10363  1.57979           3   
1           1981      1.9630    6.64537         5.78447  1.45440           3   
2           1998      2.4859    7.45601         7.23994  1.18227           2   

   MaxPlayers  ...  Rank:partygames  Rank:childrensgames  Cat:Thematic  \
0           5  ...            21926                21926             0   
1           4  ...            21926                21926             0   
2           4  ...            21926                21926             0   

  Cat:Strategy  Cat:War  Cat:Family  Cat:CGS  Cat:Abstract  Cat:Party  \
0            1        0           0        0             0          0   
1            1        0           0        0             0          0   
2            1        0           0        0             0          0   

   Cat:Childrens  
0              0  
1              0  
2              0  

[3 rows x 48 columns]

(21925, 48)


### <b> Creating a dataframe for Categories </b>

In [4]:
# Let's create a new df_categories based on the categories from games_df:
selected_columns = ['BGGId', 'Cat:Thematic', 'Cat:Strategy', 'Cat:War', 'Cat:Family', 'Cat:CGS', 'Cat:Abstract', 'Cat:Party', 'Cat:Childrens']

# Creating a new DataFrame with only the selected columns
df_categories = df_games[selected_columns].copy()

# Check:
display(df_categories)
print(df_categories.shape)

BGGId  Cat:Thematic  Cat:Strategy  Cat:War  Cat:Family  Cat:CGS  \
0           1             0             1        0           0        0   
1           2             0             1        0           0        0   
2           3             0             1        0           0        0   
3           4             0             0        0           0        0   
4           5             0             1        0           0        0   
...       ...           ...           ...      ...         ...      ...   
21920  347146             0             0        0           0        0   
21921  347521             0             0        0           0        0   
21922  348955             0             0        0           0        0   
21923  349131             0             0        0           0        0   
21924  349161             0             0        0           0        0   

       Cat:Abstract  Cat:Party  Cat:Childrens  
0                 0          0              0  
1                 0          0              0  
2                 0          0              0  
3                 0          0              0  
4                 0          0              0  
...             ...        ...            ...  
21920             0          0              0  
21921             0          0              0  
21922             0          0              0  
21923             0          0              0  
21924             0          0              0  

[21925 rows x 9 columns]

(21925, 9)


In [5]:
# Let's now pivot this table:

# Setting the index of the DataFrame to the column named 'BGGId
df_categories = df_categories.set_index('BGGId')

# Stacking the DataFrame, essentially "melting" it to convert columns into rows, 
# then resets the index, resulting in a DataFrame with three columns: 
# 'BGGId', the name of the column that was stacked, and the values.
df_categories = df_categories.stack().reset_index()

# Renaming the columns of the DataFrame to 'bgg_id', 'mechanism', and 'flag', respectively
df_categories.columns = ['BGGId', 'category', 'flag']

#  Filtering the DataFrame to keep only the rows where the 'flag' column has a value of 1, 
# and then reseting the index, dropping the old index.
df_categories = df_categories[df_categories['flag'] == 1].reset_index(drop=True)

# Droping the 'flag' column from the DataFrame, as it's no longer needed after filtering
df_categories = df_categories.drop(columns='flag')

In [6]:
# Final result after pivoting the df:
print(df_categories)
print(df_categories.shape)

        BGGId      category
0           1  Cat:Strategy
1           2  Cat:Strategy
2           3  Cat:Strategy
3           5  Cat:Strategy
4           7  Cat:Abstract
...       ...           ...
12323  339214    Cat:Family
12324  340466  Cat:Thematic
12325  342942  Cat:Strategy
12326  343562  Cat:Strategy
12327  346703    Cat:Family

[12328 rows x 2 columns]
(12328, 2)


In [7]:
# Remove 'Cat:' prefix from the 'category' column
df_categories['category'] = df_categories['category'].str.replace('Cat:', '')

# Print the modified DataFrame
display(df_categories)
print(df_categories.shape)

        BGGId  category
0           1  Strategy
1           2  Strategy
2           3  Strategy
3           5  Strategy
4           7  Abstract
...       ...       ...
12323  339214    Family
12324  340466  Thematic
12325  342942  Strategy
12326  343562  Strategy
12327  346703    Family

[12328 rows x 2 columns]
(12328, 2)


### <b> Creating a dataframe for Ranking </b>

In [8]:
# Let's create a new df_categories based on the categories from games_df:
selected_columns = ['BGGId', 'Rank:boardgame', 'Rank:strategygames', 'Rank:abstracts', 'Rank:familygames', 'Rank:thematic', 'Rank:cgs','Rank:wargames', 'Rank:partygames', 'Rank:childrensgames']

# Creating a new DataFrame with only the selected columns
df_rankings = df_games[selected_columns].copy()

# Check:
display(df_rankings)
print(df_rankings.shape)

BGGId  Rank:boardgame  Rank:strategygames  Rank:abstracts  \
0           1             316                 180           21926   
1           2            3993                1577           21926   
2           3             224                 166           21926   
3           4            5345               21926           21926   
4           5             290                 220           21926   
...       ...             ...                 ...             ...   
21920  347146           13730               21926           21926   
21921  347521           21926               21926           21926   
21922  348955           11507               21926           21926   
21923  349131           13460               21926           21926   
21924  349161           16615               21926           21926   

       Rank:familygames  Rank:thematic  Rank:cgs  Rank:wargames  \
0                 21926          21926     21926          21926   
1                 21926          21926     21926          21926   
2                 21926          21926     21926          21926   
3                 21926          21926     21926          21926   
4                 21926          21926     21926          21926   
...                 ...            ...       ...            ...   
21920             21926          21926     21926          21926   
21921             21926          21926     21926          21926   
21922             21926          21926     21926          21926   
21923             21926          21926     21926          21926   
21924             21926          21926     21926          21926   

       Rank:partygames  Rank:childrensgames  
0                21926                21926  
1                21926                21926  
2                21926                21926  
3                21926                21926  
4                21926                21926  
...                ...                  ...  
21920            21926                21926  
21921            21926                21926  
21922            21926                21926  
21923            21926                21926  
21924            21926                21926  

[21925 rows x 10 columns]

(21925, 10)


<b>Changing column names</b> : <br>
Let's now clean the column names (df_rankings header):

In [9]:
#Now, let's rename our column names:

# Get the list of column names
col_names = list(df_rankings.columns)

# Iterate through each column name
for i, name in enumerate(col_names):
    # Check if the column name starts with 'Rank:'
    if name.startswith('Rank:'):
        # Split the column name by ':' and take the second part
        new_name = name.split(':')[1]
        # Update the column name in the list
        col_names[i] = new_name

# Update the column names in the DataFrame
df_rankings.columns = col_names

In [10]:
# Print the modified DataFrame
display(df_rankings)

BGGId  boardgame  strategygames  abstracts  familygames  thematic  \
0           1        316            180      21926        21926     21926   
1           2       3993           1577      21926        21926     21926   
2           3        224            166      21926        21926     21926   
3           4       5345          21926      21926        21926     21926   
4           5        290            220      21926        21926     21926   
...       ...        ...            ...        ...          ...       ...   
21920  347146      13730          21926      21926        21926     21926   
21921  347521      21926          21926      21926        21926     21926   
21922  348955      11507          21926      21926        21926     21926   
21923  349131      13460          21926      21926        21926     21926   
21924  349161      16615          21926      21926        21926     21926   

         cgs  wargames  partygames  childrensgames  
0      21926     21926       21926           21926  
1      21926     21926       21926           21926  
2      21926     21926       21926           21926  
3      21926     21926       21926           21926  
4      21926     21926       21926           21926  
...      ...       ...         ...             ...  
21920  21926     21926       21926           21926  
21921  21926     21926       21926           21926  
21922  21926     21926       21926           21926  
21923  21926     21926       21926           21926  
21924  21926     21926       21926           21926  

[21925 rows x 10 columns]

In [11]:
# Let's now add the '_rank' at the end of each column (except 'BGGId') 

# Iterate through each column name, excluding the first column
for i, name in enumerate(col_names[1:], start=1):
    # Append '_rank' to each column name
    new_name = name + '_rank'
    # Update the column name in the list
    col_names[i] = new_name

# Update the column names in the DataFrame
df_rankings.columns = col_names

# Print the modified DataFrame
display(df_rankings)

BGGId  boardgame_rank  strategygames_rank  abstracts_rank  \
0           1             316                 180           21926   
1           2            3993                1577           21926   
2           3             224                 166           21926   
3           4            5345               21926           21926   
4           5             290                 220           21926   
...       ...             ...                 ...             ...   
21920  347146           13730               21926           21926   
21921  347521           21926               21926           21926   
21922  348955           11507               21926           21926   
21923  349131           13460               21926           21926   
21924  349161           16615               21926           21926   

       familygames_rank  thematic_rank  cgs_rank  wargames_rank  \
0                 21926          21926     21926          21926   
1                 21926          21926     21926          21926   
2                 21926          21926     21926          21926   
3                 21926          21926     21926          21926   
4                 21926          21926     21926          21926   
...                 ...            ...       ...            ...   
21920             21926          21926     21926          21926   
21921             21926          21926     21926          21926   
21922             21926          21926     21926          21926   
21923             21926          21926     21926          21926   
21924             21926          21926     21926          21926   

       partygames_rank  childrensgames_rank  
0                21926                21926  
1                21926                21926  
2                21926                21926  
3                21926                21926  
4                21926                21926  
...                ...                  ...  
21920            21926                21926  
21921            21926                21926  
21922            21926                21926  
21923            21926                21926  
21924            21926                21926  

[21925 rows x 10 columns]

In [12]:
print(df_rankings.dtypes)

BGGId                  int64
boardgame_rank         int64
strategygames_rank     int64
abstracts_rank         int64
familygames_rank       int64
thematic_rank          int64
cgs_rank               int64
wargames_rank          int64
partygames_rank        int64
childrensgames_rank    int64
dtype: object


<b> testing things, don't mind it </b>

In [13]:
'''
for i, name in enumerate(col_names):
    col_names[i] = name.split(':')

col_names
'''

"\nfor i, name in enumerate(col_names):\n    col_names[i] = name.split(':')\n\ncol_names\n"

### <b> Mechanics </b>

In [14]:
# Importing the main dataframe 'mechanics.csv'
df_mechanics = pd.read_csv(file_path + '/mechanics.csv')

# Quick view:
df_mechanics.head(3)

BGGId  Alliances  Area Majority / Influence  Auction/Bidding  Dice Rolling  \
0      1          1                          1                1             1   
1      2          0                          0                0             0   
2      3          0                          1                0             0   

   Hand Management  Simultaneous Action Selection  Trick-taking  Hexagon Grid  \
0                1                              1             0             0   
1                0                              0             1             0   
2                1                              0             0             1   

   Once-Per-Game Abilities  ...  Contracts  Passed Action Token  \
0                        0  ...          0                    0   
1                        0  ...          0                    0   
2                        1  ...          0                    0   

   King of the Hill  Action Retrieval  Force Commitment  Rondel  \
0                 0                 0                 0       0   
1                 0                 0                 0       0   
2                 0                 0                 0       0   

   Automatic Resource Growth  Legacy Game  Dexterity  Physical  
0                          0            0          0         0  
1                          0            0          0         0  
2                          0            0          0         0  

[3 rows x 158 columns]

In [15]:
# Setting the index of the DataFrame to the column named 'BGGId
df_mechanics = df_mechanics.set_index('BGGId')

# Stacking the DataFrame, essentially "melting" it to convert columns into rows, 
# then resets the index, resulting in a DataFrame with three columns: 
# 'BGGId', the name of the column that was stacked, and the values.
df_mechanics = df_mechanics.stack().reset_index()

# Renaming the columns of the DataFrame to 'bgg_id', 'mechanism', and 'flag', respectively
df_mechanics.columns = ['bgg_id', 'mechanism', 'flag']

#  Filtering the DataFrame to keep only the rows where the 'flag' column has a value of 1, 
# and then reseting the index, dropping the old index.
df_mechanics = df_mechanics[df_mechanics['flag'] == 1].reset_index(drop=True)

# Droping the 'flag' column from the DataFrame, as it's no longer needed after filtering
df_mechanics = df_mechanics.drop(columns='flag')

In [64]:
# Final result after pivoting the df:
display(df_mechanics)
print(df_mechanics.shape)

bgg_id                  mechanism
0           1                  Alliances
1           1  Area Majority / Influence
2           1            Auction/Bidding
3           1               Dice Rolling
4           1            Hand Management
...       ...                        ...
68075  349131           Pattern Building
68076  349161               Dice Rolling
68077  349161           Paper-and-Pencil
68078  349161      Solo / Solitaire Game
68079  349161           End Game Bonuses

[68080 rows x 2 columns]

(68080, 2)


### <b> Themes </b>

In [17]:
# Importing the main dataframe 'themes.csv'
df_themes = pd.read_csv(file_path + '/themes.csv')

# Quick view:
df_themes.head(3)

BGGId  Adventure  Fantasy  Fighting  Environmental  Medical  Economic  \
0      1          0        0         0              0        0         1   
1      2          0        1         0              0        0         0   
2      3          0        0         0              0        0         0   

   Industry / Manufacturing  Transportation  Science Fiction  ...  \
0                         0               0                0  ...   
1                         0               0                0  ...   
2                         0               0                0  ...   

   Theme_Fashion  Theme_Geocaching  Theme_Ecology  Theme_Chernobyl  \
0              0                 0              0                0   
1              0                 0              0                0   
2              0                 0              0                0   

   Theme_Photography  Theme_French Foreign Legion  Theme_Cruise ships  \
0                  0                            0                   0   
1                  0                            0                   0   
2                  0                            0                   0   

   Theme_Apache Tribes  Theme_Rivers  Theme_Flags identification  
0                    0             0                           0  
1                    0             0                           0  
2                    0             0                           0  

[3 rows x 218 columns]

In [18]:
# Setting the index of the DataFrame to the column named 'BGGId
df_themes = df_themes.set_index('BGGId')

# Stacking the DataFrame, essentially "melting" it to convert columns into rows, 
# then resets the index, resulting in a DataFrame with three columns: 
# 'BGGId', the name of the column that was stacked, and the values.
df_themes = df_themes.stack().reset_index()

# Renaming the columns of the DataFrame to 'bgg_id', 'mechanism', and 'flag', respectively
df_themes.columns = ['bgg_id', 'theme', 'flag']

#  Filtering the DataFrame to keep only the rows where the 'flag' column has a value of 1, 
# and then reseting the index, dropping the old index.
df_themes = df_themes[df_themes['flag'] == 1].reset_index(drop=True)

# Droping the 'flag' column from the DataFrame, as it's no longer needed after filtering
df_themes = df_themes.drop(columns='flag')

In [65]:
# Final result after pivoting the df:
display(df_themes)
print(df_themes.shape)

bgg_id          theme
0           1       Economic
1           1      Political
2           2        Fantasy
3           3       Medieval
4           3  Theme_Samurai
...       ...            ...
32374  346965    Renaissance
32375  346965      Theme_Art
32376  347521   World War II
32377  349161       Nautical
32378  349161        Pirates

[32379 rows x 2 columns]

(32379, 2)


### <b> Subcategories </b>

In [20]:
# Importing the main dataframe 'subcategories.csv'
df_subcategories = pd.read_csv(file_path + '/subcategories.csv')

# Quick view:
df_subcategories.head(3)

BGGId  Exploration  Miniatures  Territory Building  Card Game  Educational  \
0      1            0           0                   0          0            0   
1      2            0           0                   0          1            0   
2      3            0           0                   0          0            0   

   Puzzle  Collectible Components  Word Game  Print & Play  Electronic  
0       0                       0          0             0           0  
1       0                       0          0             0           0  
2       0                       0          0             0           0

In [21]:
# Setting the index of the DataFrame to the column named 'BGGId
df_subcategories = df_subcategories.set_index('BGGId')

# Stacking the DataFrame, essentially "melting" it to convert columns into rows, 
# then resets the index, resulting in a DataFrame with three columns: 
# 'BGGId', the name of the column that was stacked, and the values.
df_subcategories = df_subcategories.stack().reset_index()

# Renaming the columns of the DataFrame to 'bgg_id', 'mechanism', and 'flag', respectively
df_subcategories.columns = ['bgg_id', 'subcategory', 'flag']

#  Filtering the DataFrame to keep only the rows where the 'flag' column has a value of 1, 
# and then reseting the index, dropping the old index.
df_subcategories = df_subcategories[df_subcategories['flag'] == 1].reset_index(drop=True)

# Droping the 'flag' column from the DataFrame, as it's no longer needed after filtering
df_subcategories = df_subcategories.drop(columns='flag')

In [66]:
# Final result after pivoting the df:
display(df_subcategories)
print(df_subcategories.shape)

bgg_id             subcategory
0           2               Card Game
1           5      Territory Building
2           9             Exploration
3          11               Card Game
4          17  Collectible Components
...       ...                     ...
11805  345976               Card Game
11806  346482                  Puzzle
11807  346965                  Puzzle
11808  347146               Card Game
11809  348955            Print & Play

[11810 rows x 2 columns]

(11810, 2)


### <b> Artists (reduced)</b>

In [23]:
# Importing the main dataframe 'artists_reduced.csv'
df_artists = pd.read_csv(file_path + '/artists_reduced.csv')

# Quick view:
df_artists.head(3)

Harald Lieske  Franz Vohwinkel  Peter Whitley  Scott Okumura  (Uncredited)  \
0              1                0              0              0             0   
1              0                0              0              0             0   
2              0                1              0              0             0   

   Doris Matthäus  Alan R. Moon  Alexander Jung  Andrea Boekhoff  \
0               0             0               0                0   
1               0             0               0                0   
2               0             0               0                0   

   Björn Pertoft  ...  Nathan Meunier  Andrey Gordeev  Zbigniew Umgelter  \
0              0  ...               0               0                  0   
1              0  ...               0               0                  0   
2              0  ...               0               0                  0   

   Jeppe Norsker  Daniel Profiri  Aleksander Zawada  Simon Douchy  \
0              0               0                  0             0   
1              0               0                  0             0   
2              0               0                  0             0   

   Felix Wermke  BGGId  Low-Exp Artist  
0             0      1               1  
1             0      2               1  
2             0      3               0  

[3 rows x 1681 columns]

In [24]:
# Setting the index of the DataFrame to the column named 'BGGId
df_artists = df_artists.set_index('BGGId')

# Stacking the DataFrame, essentially "melting" it to convert columns into rows, 
# then resets the index, resulting in a DataFrame with three columns: 
# 'BGGId', the name of the column that was stacked, and the values.
df_artists = df_artists.stack().reset_index()

# Renaming the columns of the DataFrame to 'bgg_id', 'mechanism', and 'flag', respectively
df_artists.columns = ['BGGId', 'artist', 'flag']

#  Filtering the DataFrame to keep only the rows where the 'flag' column has a value of 1, 
# and then reseting the index, dropping the old index.
df_artists = df_artists[df_artists['flag'] == 1].reset_index(drop=True)

# Droping the 'flag' column from the DataFrame, as it's no longer needed after filtering
df_artists = df_artists.drop(columns='flag')

In [67]:
# Final result after pivoting the df:
display(df_artists)
print(df_artists.shape)

BGGId               artist
0           1        Harald Lieske
1           1       Low-Exp Artist
2           2       Low-Exp Artist
3           3      Franz Vohwinkel
4           4       Low-Exp Artist
...       ...                  ...
26163  347521         Alan D'Amico
26164  347521       Low-Exp Artist
26165  348955           Beth Sobel
26166  349131  Oliver Freudenreich
26167  349161  Olivier Derouetteau

[26168 rows x 2 columns]

(26168, 2)


### <b> Designers (reduced) </b>

In [26]:
# Importing the main dataframe 'designers_reduced.csv'
df_designers = pd.read_csv(file_path + '/designers_reduced.csv')

# Quick view:
df_designers.head(3)

Karl-Heinz Schmiel  G. W. "Jerry" D'Arcey  Reiner Knizia  Sid Sackson  \
0                   1                      0              0            0   
1                   0                      1              0            0   
2                   0                      0              1            0   

   Jean du Poël  Martin Wallace  Richard Ulrich  Wolfgang Kramer  \
0             0               0               0                0   
1             0               0               0                0   
2             0               0               0                0   

   Alan R. Moon  Uwe Rosenberg  ...  Thomas Dupont  Mathieu Casnin  \
0             0              0  ...              0               0   
1             0              0  ...              0               0   
2             0              0  ...              0               0   

   Sean Fletcher  Moritz Dressler  Molly Johnson  Robert Melvin  \
0              0                0              0              0   
1              0                0              0              0   
2              0                0              0              0   

   Shawn Stankewich  Nathan Meunier  BGGId  Low-Exp Designer  
0                 0               0      1                 0  
1                 0               0      2                 0  
2                 0               0      3                 0  

[3 rows x 1594 columns]

In [27]:
# Setting the index of the DataFrame to the column named 'BGGId
df_designers = df_designers.set_index('BGGId')

# Stacking the DataFrame, essentially "melting" it to convert columns into rows, 
# then resets the index, resulting in a DataFrame with three columns: 
# 'BGGId', the name of the column that was stacked, and the values.
df_designers = df_designers.stack().reset_index()

# Renaming the columns of the DataFrame to 'bgg_id', 'mechanism', and 'flag', respectively
df_designers.columns = ['BGGId', 'designer', 'flag']

#  Filtering the DataFrame to keep only the rows where the 'flag' column has a value of 1, 
# and then reseting the index, dropping the old index.
df_designers = df_designers[df_designers['flag'] == 1].reset_index(drop=True)

# Droping the 'flag' column from the DataFrame, as it's no longer needed after filtering
df_designers = df_designers.drop(columns='flag')

In [68]:
# Final result after pivoting the df:
display(df_designers)
print(df_designers.shape)

BGGId               designer
0           1     Karl-Heinz Schmiel
1           2  G. W. "Jerry" D'Arcey
2           3          Reiner Knizia
3           4       Low-Exp Designer
4           5            Sid Sackson
...       ...                    ...
26738  347146       Low-Exp Designer
26739  347521             Paolo Mori
26740  348955       Low-Exp Designer
26741  349131       Low-Exp Designer
26742  349161       Low-Exp Designer

[26743 rows x 2 columns]

(26743, 2)


### <b> Publishers (reduced) </b>

In [29]:
# Importing the main dataframe 'publishers_reduced.csv'
df_publishers = pd.read_csv(file_path + '/publishers_reduced.csv')

# Quick view:
df_publishers.head(3)

Hans im Glück  Moskito Spiele  Portal Games  Spielworxx  Stronghold Games  \
0              1               1             1           1                 1   
1              0               0             0           0                 0   
2              1               0             0           0                 0   

   Valley Games, Inc.  YOKA Games  sternenschimmermeer  E.S. Lowe  \
0                   1           1                    1          0   
1                   0           0                    0          1   
2                   0           0                    0          0   

   Milton Bradley  ...  Cacahuete Games  BlackSands Games  Norsker Games  \
0               0  ...                0                 0              0   
1               1  ...                0                 0              0   
2               0  ...                0                 0              0   

   Perro Loko Games  Funko Games  Origame  Deep Print Games  \
0                 0            0        0                 0   
1                 0            0        0                 0   
2                 0            0        0                 0   

   Hidden Industries GmbH  BGGId  Low-Exp Publisher  
0                       0      1                  0  
1                       0      2                  0  
2                       0      3                  1  

[3 rows x 1866 columns]

In [30]:
# Setting the index of the DataFrame to the column named 'BGGId
df_publishers = df_publishers.set_index('BGGId')

# Stacking the DataFrame, essentially "melting" it to convert columns into rows, 
# then resets the index, resulting in a DataFrame with three columns: 
# 'BGGId', the name of the column that was stacked, and the values.
df_publishers = df_publishers.stack().reset_index()

# Renaming the columns of the DataFrame to 'bgg_id', 'mechanism', and 'flag', respectively
df_publishers.columns = ['BGGId', 'publishing_company', 'flag']

#  Filtering the DataFrame to keep only the rows where the 'flag' column has a value of 1, 
# and then reseting the index, dropping the old index.
df_publishers = df_publishers[df_publishers['flag'] == 1].reset_index(drop=True)

# Droping the 'flag' column from the DataFrame, as it's no longer needed after filtering
df_publishers = df_publishers.drop(columns='flag')

In [69]:
# Final result after pivoting the df:
display(df_publishers)
print(df_publishers.shape)

BGGId             publishing_company
0           1                  Hans im Glück
1           1                 Moskito Spiele
2           1                   Portal Games
3           1                     Spielworxx
4           1               Stronghold Games
...       ...                            ...
56604  347146                  La Mame Games
56605  347521                      PSC Games
56606  348955               (Self-Published)
56607  349131  Nürnberger-Spielkarten-Verlag
56608  349161               Bragelonne Games

[56609 rows x 2 columns]

(56609, 2)


### <b> User Ratings </b>

Since that csv is above 1 million rows, i will have to skip it cause I cannot run the cell  below - fortunately I don't really need that csv

In [32]:
# Importing the main dataframe 'user_ratings.csv'
df_user_ratings = pd.read_csv(file_path + '/user_ratings.csv')

# Display the dataframe:
display(df_user_ratings.head(3))

# looking at the size of the df
print(df_user_ratings.shape)

BGGId  Rating    Username
0  213788     8.0   Tonydorrf
1  213788     8.0  tachyon14k
2  213788     8.0    Ungotter

(18942215, 3)


### <b> Ratings Distribution </b>

In [33]:
# Importing the main dataframe 'ratings_distribution.csv'
df_ratings_distribution = pd.read_csv(file_path + '/ratings_distribution.csv')

# Display the dataframe:
display(df_ratings_distribution.head(3))

# looking at the size of the df
print(df_ratings_distribution.shape)

BGGId  0.0  0.1  0.5  1.0  1.1  1.2  1.3  1.4  1.5  ...   9.2  9.3   9.4  \
0      1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  11.0  5.0  11.0   
1      2  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  ...   0.0  1.0   0.0   
2      3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  20.0  7.0   4.0   

    9.5  9.6  9.7  9.8  9.9   10.0  total_ratings  
0  86.0  3.0  4.0  6.0  8.0  426.0         5352.0  
1   1.0  0.0  0.0  1.0  0.0   17.0          562.0  
2  77.0  3.0  1.0  5.0  3.0  477.0        15148.0  

[3 rows x 96 columns]

(21925, 96)


In [34]:
# Checking for a column data type:

# Using dtype attribute
print(df_ratings_distribution['total_ratings'].dtype)

# Using dtypes property
print(df_ratings_distribution.dtypes['total_ratings'])

# for each column
print(df_ratings_distribution.dtypes)

float64
float64
BGGId              int64
0.0              float64
0.1              float64
0.5              float64
1.0              float64
                  ...   
9.7              float64
9.8              float64
9.9              float64
10.0             float64
total_ratings    float64
Length: 96, dtype: object


<b> Converting the float columns to integers </b>

In [35]:
# Convert float columns to integers
df_ratings_distribution.iloc[:, 1:] = df_ratings_distribution.iloc[:, 1:].astype(int)

# check the data types:
print(df_ratings_distribution.columns.dtype)

# Display the converted DataFrame
display(df_ratings_distribution.head(3))

object


BGGId  0.0  0.1  0.5  1.0  1.1  1.2  1.3  1.4  1.5  ...   9.2  9.3   9.4  \
0      1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  11.0  5.0  11.0   
1      2  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  ...   0.0  1.0   0.0   
2      3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  20.0  7.0   4.0   

    9.5  9.6  9.7  9.8  9.9   10.0  total_ratings  
0  86.0  3.0  4.0  6.0  8.0  426.0         5352.0  
1   1.0  0.0  0.0  1.0  0.0   17.0          562.0  
2  77.0  3.0  1.0  5.0  3.0  477.0        15148.0  

[3 rows x 96 columns]

Not working here - could work if we looped. <br>
--> We'll use the <i> applymap <i> function instead.

In [36]:
#transformin all columns to integer
df_ratings_distribution = df_ratings_distribution.applymap(int)

#looking at the final result
df_ratings_distribution.head()

C:\Users\micel\AppData\Local\Temp\ipykernel_9496\4285546565.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_ratings_distribution = df_ratings_distribution.applymap(int)


BGGId  0.0  0.1  0.5  1.0  1.1  1.2  1.3  1.4  1.5  ...  9.2  9.3  9.4  \
0      1    0    0    0    0    0    0    0    0    0  ...   11    5   11   
1      2    0    0    0    3    0    0    0    0    0  ...    0    1    0   
2      3    0    0    0    0    0    0    0    0    0  ...   20    7    4   
3      4    0    0    0    0    0    0    0    0    0  ...    0    0    0   
4      5    0    0    0    0    0    0    0    0    0  ...   18    8    4   

   9.5  9.6  9.7  9.8  9.9  10.0  total_ratings  
0   86    3    4    6    8   426           5352  
1    1    0    0    1    0    17            562  
2   77    3    1    5    3   477          15148  
3    0    0    0    0    0     3            342  
4   82    7    9   10    5   905          18387  

[5 rows x 96 columns]

In [37]:
# checking with dtypes
df_ratings_distribution.dtypes

BGGId            int64
0.0              int64
0.1              int64
0.5              int64
1.0              int64
                 ...  
9.7              int64
9.8              int64
9.9              int64
10.0             int64
total_ratings    int64
Length: 96, dtype: object

# EDA: Exploring and Analyzing

<b>Precision:</b>
Since all the datasets use the same primary key (BGGId) and that my main and most important dataframe is df_games, I will focus my EDA on it.
It currently has 21925 rows like the other dataframes, the primary_key being the same all the time (BGGId) and 48 columns !

## Data Types, info, describe, shape

In [38]:
# Shape of the dataframe
df_games.shape

(21925, 48)

In [39]:
# columns data types 
df_games.dtypes

BGGId                    int64
Name                    object
Description             object
YearPublished            int64
GameWeight             float64
AvgRating              float64
BayesAvgRating         float64
StdDev                 float64
MinPlayers               int64
MaxPlayers               int64
ComAgeRec              float64
LanguageEase           float64
BestPlayers              int64
GoodPlayers             object
NumOwned                 int64
NumWant                  int64
NumWish                  int64
NumWeightVotes           int64
MfgPlaytime              int64
ComMinPlaytime           int64
ComMaxPlaytime           int64
MfgAgeRec                int64
NumUserRatings           int64
NumComments              int64
NumAlternates            int64
NumExpansions            int64
NumImplementations       int64
IsReimplementation       int64
Family                  object
Kickstarted              int64
ImagePath               object
Rank:boardgame           int64
Rank:str

In [40]:
#quick view of the stats
df_games.describe()

BGGId  YearPublished    GameWeight     AvgRating  \
count   21925.000000   21925.000000  21925.000000  21925.000000   
mean   117652.663216    1985.494914      1.982131      6.424922   
std    104628.721777     212.486214      0.848983      0.932477   
min         1.000000   -3500.000000      0.000000      1.041330   
25%     12346.000000    2001.000000      1.333300      5.836960   
50%    105305.000000    2011.000000      1.968800      6.453950   
75%    206169.000000    2017.000000      2.525200      7.052450   
max    349161.000000    2021.000000      5.000000      9.914290   

       BayesAvgRating        StdDev    MinPlayers    MaxPlayers     ComAgeRec  \
count    21925.000000  21925.000000  21925.000000  21925.000000  16395.000000   
mean         5.685673      1.516374      2.007343      5.707868     10.004391   
std          0.365311      0.285578      0.693093     15.014643      3.269157   
min          3.574810      0.196023      0.000000      0.000000      2.000000   
25%          5.510300      1.320720      2.000000      4.000000      8.000000   
50%          5.546540      1.476880      2.000000      4.000000     10.000000   
75%          5.679890      1.665470      2.000000      6.000000     12.000000   
max          8.514880      4.277280     10.000000    999.000000     21.000000   

       LanguageEase  ...  Rank:partygames  Rank:childrensgames  Cat:Thematic  \
count  16034.000000  ...     21925.000000         21925.000000  21925.000000   
mean     216.461819  ...     21295.352201         21062.680274      0.055827   
std      236.595136  ...      3637.139987          4219.776597      0.229592   
min        1.000000  ...         1.000000             1.000000      0.000000   
25%       24.027778  ...     21926.000000         21926.000000      0.000000   
50%      138.000000  ...     21926.000000         21926.000000      0.000000   
75%      351.000000  ...     21926.000000         21926.000000      0.000000   
max     1757.000000  ...     21926.000000         21926.000000      1.000000   

       Cat:Strategy       Cat:War    Cat:Family       Cat:CGS  Cat:Abstract  \
count   21925.00000  21925.000000  21925.000000  21925.000000  21925.000000   
mean        0.10577      0.161003      0.105633      0.013820      0.050855   
std         0.30755      0.367542      0.307374      0.116745      0.219707   
min         0.00000      0.000000      0.000000      0.000000      0.000000   
25%         0.00000      0.000000      0.000000      0.000000      0.000000   
50%         0.00000      0.000000      0.000000      0.000000      0.000000   
75%         0.00000      0.000000      0.000000      0.000000      0.000000   
max         1.00000      1.000000      1.000000      1.000000      1.000000   

          Cat:Party  Cat:Childrens  
count  21925.000000   21925.000000  
mean       0.029190       0.040182  
std        0.168344       0.196391  
min        0.000000       0.000000  
25%        0.000000       0.000000  
50%        0.000000       0.000000  
75%        0.000000       0.000000  
max        1.000000       1.000000  

[8 rows x 43 columns]

In [41]:
#info about the dataframe
df_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21925 entries, 0 to 21924
Data columns (total 48 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   BGGId                21925 non-null  int64  
 1   Name                 21925 non-null  object 
 2   Description          21924 non-null  object 
 3   YearPublished        21925 non-null  int64  
 4   GameWeight           21925 non-null  float64
 5   AvgRating            21925 non-null  float64
 6   BayesAvgRating       21925 non-null  float64
 7   StdDev               21925 non-null  float64
 8   MinPlayers           21925 non-null  int64  
 9   MaxPlayers           21925 non-null  int64  
 10  ComAgeRec            16395 non-null  float64
 11  LanguageEase         16034 non-null  float64
 12  BestPlayers          21925 non-null  int64  
 13  GoodPlayers          21925 non-null  object 
 14  NumOwned             21925 non-null  int64  
 15  NumWant              21925 non-null 

## Null values

### What columns have Null values ?

In [42]:
# Checking for columns containing at least one null value
mask = df_games.isna().any()
df_with_null = df_games.loc[:, mask]

df_with_null.columns

Index(['Description', 'ComAgeRec', 'LanguageEase', 'Family', 'ImagePath'], dtype='object')

'Description', 'ComAgeRec', 'LanguageEase', 'Family', 'ImagePath' = the columns having at least one NaN

In [43]:
df_games.isna().any(axis=1)

0        False
1         True
2        False
3         True
4        False
         ...  
21920     True
21921     True
21922     True
21923     True
21924     True
Length: 21925, dtype: bool

In [44]:
# Checking for null values
mask = df_games.isna().any()
columns_with_null = df_games.columns[mask]

# Loop through columns with null values and calculate the count of null values for each column
for column in columns_with_null:
    null_count = df_games[column].isna().sum()
    print(f"'{column}' : {null_count} missing values.")

'Description' : 1 missing values.
'ComAgeRec' : 5530 missing values.
'LanguageEase' : 5891 missing values.
'Family' : 15262 missing values.
'ImagePath' : 17 missing values.


In [45]:
# Checking for null values
mask = df_games.isna().any()
columns_with_null = df_games.loc[:, mask]

# Calculate the quantity of missing values for each column
missing_values_count = columns_with_null.isna().sum()

print("Missing values count for columns with at least one null value:")
print(missing_values_count)

Missing values count for columns with at least one null value:
Description         1
ComAgeRec        5530
LanguageEase     5891
Family          15262
ImagePath          17
dtype: int64


### Getting the proportion of those missing values

In [46]:
# Calculate the total number of rows in the DataFrame
total_rows = len(df_games)

# Calculate the proportion of missing values for each column
proportion_missing_values = (missing_values_count / total_rows) * 100

print("Proportion of missing values for columns with at least one null value:")
print(proportion_missing_values)

Proportion of missing values for columns with at least one null value:
Description      0.004561
ComAgeRec       25.222349
LanguageEase    26.868871
Family          69.610034
ImagePath        0.077537
dtype: float64


- 25% of the rows are missing the minimmum age recommended
- 26%
- 69%


## Duplicates

In [47]:
duplicate_rows = df_games[df_games.duplicated()]
num_duplicate_rows = duplicate_rows.shape[0]

if num_duplicate_rows > 0:
    print("Number of duplicate rows:", num_duplicate_rows)
    print("Duplicate rows:")
    print(duplicate_rows)
else:
    print("No duplicate rows found.")

No duplicate rows found.


## Unique Values ?

## Check for Outliers

## Various Viz

# Cleaning

## df_games: droping the categories and ranking columns

In [48]:
# Dictionary containing DataFrame names and their corresponding DataFrames
dataframes = {
    'games': df_games,
    # (Add more DataFrames here if needed)
}

In [49]:
# Iterate through each DataFrame and print column names and data types

for name, df in dataframes.items():
    print(f"DataFrame Name: {name}")
    print("Column Names and Data Types:")
    for col_name, dtype in df.dtypes.items():
        print(f"- {col_name}: {dtype}")
    print("\n")

DataFrame Name: games
Column Names and Data Types:
- BGGId: int64
- Name: object
- Description: object
- YearPublished: int64
- GameWeight: float64
- AvgRating: float64
- BayesAvgRating: float64
- StdDev: float64
- MinPlayers: int64
- MaxPlayers: int64
- ComAgeRec: float64
- LanguageEase: float64
- BestPlayers: int64
- GoodPlayers: object
- NumOwned: int64
- NumWant: int64
- NumWish: int64
- NumWeightVotes: int64
- MfgPlaytime: int64
- ComMinPlaytime: int64
- ComMaxPlaytime: int64
- MfgAgeRec: int64
- NumUserRatings: int64
- NumComments: int64
- NumAlternates: int64
- NumExpansions: int64
- NumImplementations: int64
- IsReimplementation: int64
- Family: object
- Kickstarted: int64
- ImagePath: object
- Rank:boardgame: int64
- Rank:strategygames: int64
- Rank:abstracts: int64
- Rank:familygames: int64
- Rank:thematic: int64
- Rank:cgs: int64
- Rank:wargames: int64
- Rank:partygames: int64
- Rank:childrensgames: int64
- Cat:Thematic: int64
- Cat:Strategy: int64
- Cat:War: int64
- Cat:Fam

In [50]:
# Now, let's drop the 'Cat:xxx' and 'Rank:xxx' columns:

# List comprehension to filter out columns not starting with 'Rank:' or 'Cat:'
columns_to_keep = [col for col in df_games.columns if not (col.startswith('Rank:') or col.startswith('Cat:'))]

# Drop the filtered columns from the DataFrame
df_games = df_games[columns_to_keep]

In [51]:
# Check the result:
display(df_games.head(2))

BGGId          Name                                        Description  \
0      1    Die Macher  die macher game seven sequential political rac...   
1      2  Dragonmaster  dragonmaster tricktaking card game base old ga...   

   YearPublished  GameWeight  AvgRating  BayesAvgRating   StdDev  MinPlayers  \
0           1986      4.3206    7.61428         7.10363  1.57979           3   
1           1981      1.9630    6.64537         5.78447  1.45440           3   

   MaxPlayers  ...  MfgAgeRec  NumUserRatings  NumComments NumAlternates  \
0           5  ...         14            5354            0             2   
1           4  ...         12             562            0             0   

   NumExpansions  NumImplementations  IsReimplementation  \
0              0                   0                   0   
1              0                   2                   1   

                        Family  Kickstarted  \
0  Classic Line (Valley Games)            0   
1                          NaN            0   

                                           ImagePath  
0  https://cf.geekdo-images.com/rpwCZAjYLD940NWwP...  
1  https://cf.geekdo-images.com/oQYhaJx5Lg3KcGis2...  

[2 rows x 31 columns]

## Remove duplicates

In [52]:
# df_games size BEFORE removal
print(df_games.shape)

# Remove duplicate rows from the DataFrame
df_games = df_games.drop_duplicates()

# df_games size AFTER removal
print(df_games.shape)

(21925, 31)
(21925, 31)


Like shown before above, there is no duplicate rows so no difference

## Handles NaNs

In [53]:
# df_games size BEFORE dropna()
print(df_games.shape)


# Remove rows that are entirely null
df_games = df_games.dropna(how='all')

# df_games size AFTER dropna()
print(df_games.shape)

(21925, 31)
(21925, 31)


## Handle date formats (year ?)

## Change columns names / formats

In [54]:
#renaming a specific column:
# predictions.rename(columns={'S.No': 'character_id'}, inplace=True)
# predictions

## Unnecessary columns to remove

## Floats to round

## Outliers hanfling / removing

In [55]:
for col_name, dtype in df_games.dtypes.items():
    print(f"'{col_name}': {dtype}")

'BGGId': int64
'Name': object
'Description': object
'YearPublished': int64
'GameWeight': float64
'AvgRating': float64
'BayesAvgRating': float64
'StdDev': float64
'MinPlayers': int64
'MaxPlayers': int64
'ComAgeRec': float64
'LanguageEase': float64
'BestPlayers': int64
'GoodPlayers': object
'NumOwned': int64
'NumWant': int64
'NumWish': int64
'NumWeightVotes': int64
'MfgPlaytime': int64
'ComMinPlaytime': int64
'ComMaxPlaytime': int64
'MfgAgeRec': int64
'NumUserRatings': int64
'NumComments': int64
'NumAlternates': int64
'NumExpansions': int64
'NumImplementations': int64
'IsReimplementation': int64
'Family': object
'Kickstarted': int64
'ImagePath': object


# Connecting to MySQL + generate the tables

## Creating the connection

In [56]:
import pandas as pd
from sqlalchemy import create_engine, text
import pymysql.cursors
import os
import getpass
import urllib.parse

<b>pymysql</b> .cursors : for connecting to MySQL database. <br>
<b>os</b> : for interacting with the operating system. <br>
<b>urllib.parse</b> : for URL encoding



In [57]:
pw_raw = 'Silver57' #os.getenv('ironhack') (--> suggests an attempt to retrieve the password from an environment variable named 'ironhack')
pw = urllib.parse.quote_plus(pw_raw)

In [58]:

# --> This line constructs a connection string for MySQL database. 
# It includes the username ('root'), encoded password (pw), host ('localhost'), port number ('3306'), 
# and the name of the database ('bgg').

connection_string = 'mysql+pymysql://root:' + pw + '@localhost:3306/bgg'


# It creates a database engine using the connection string created earlier. 
# The engine is an interface to the database that allows executing SQL queries.
engine = create_engine(connection_string)

In [59]:
# Data Insertion:
df_games.to_sql('games', engine, 'bgg', if_exists='replace', index=False)

C:\Users\micel\AppData\Local\Temp\ipykernel_9496\4223941043.py:2: FutureWarning: Starting with pandas version 3.0 all arguments of to_sql except for the arguments 'name' and 'con' will be keyword-only.
  df_games.to_sql('games', engine, 'bgg', if_exists='replace', index=False)


21925

In [60]:
pip install cryptography

Note: you may need to restart the kernel to use updated packages.


In [61]:
# df_categories.to_sql('categories', engine, 'bgg', if_exists='replace', index=False)
# df_rankings.to_sql('rankings', engine, 'bgg', if_exists='replace', index=False)
# df_mechanics.to_sql('mechanics', engine, 'bgg', if_exists='replace', index=False)
# df_themes.to_sql('themes', engine, 'bgg', if_exists='replace', index=False)
# df_subcategories.to_sql('subcategories', engine, 'bgg', if_exists='replace', index=False)
# df_artists.to_sql('artists', engine, 'bgg', if_exists='replace', index=False)
# df_designers.to_sql('designers', engine, 'bgg', if_exists='replace', index=False)
# df_publishers.to_sql('publishers', engine, 'bgg', if_exists='replace', index=False)

# --> ISSUE !
#df_user_ratings.to_sql('user_ratings', engine, 'bgg', if_exists='replace', index=False)


# df_ratings_distribution.to_sql('ratings_distribution', engine, 'bgg', if_exists='replace', index=False)

C:\Users\micel\AppData\Local\Temp\ipykernel_9496\2534227718.py:2: FutureWarning: Starting with pandas version 3.0 all arguments of to_sql except for the arguments 'name' and 'con' will be keyword-only.
  df_rankings.to_sql('rankings', engine, 'bgg', if_exists='replace', index=False)


21925

The following should be working if we forget about the df_user_ratings:

In [63]:
# Looping to repeat the process for all the other dataframes

#listing all the dataframes:
dataframes = {
    'categories': df_categories,
    'mechanics': df_mechanics,
    'themes': df_themes,
    'subcategories' : df_subcategories,
    'rankings' : df_rankings,
    'artists' : df_artists,
    'designers' : df_designers,
    'publishers' : df_publishers,
    #'user_ratings': df_user_ratings,
    'ratings_distribution' : df_ratings_distribution
}

# Looping to send them all as tables on MySQL:
for table_name, df in dataframes.items():
#    pw_raw = 'Silver57' #os.getenv('ironhack') (--> suggests an attempt to retrieve the password from an environment variable named 'ironhack')
#    pw = urllib.parse.quote_plus(pw_raw)
#    connection_string = 'mysql+pymysql://root:' + pw + '@localhost:3306/bgg'
#    engine = create_engine(connection_string)
    
    df.to_sql(table_name, engine, 'bgg', if_exists='replace', index=False)

C:\Users\micel\AppData\Local\Temp\ipykernel_9496\2176568556.py:24: FutureWarning: Starting with pandas version 3.0 all arguments of to_sql except for the arguments 'name' and 'con' will be keyword-only.
  df.to_sql(table_name, engine, 'bgg', if_exists='replace', index=False)
